### ISSUE TITLE
[request] Merge animation data + Isolation export of a tracks/layers

### ISSUE THREAD

**Issue by bencresty** [EsotericSoftware/spine-editor#665]
1) Isolation export

When creating character animations for video I prefer to use Spine.
Sometimes in the workflow there are other tools involved tho to composite afterwards. Like in this case Cavalry, After Effects or Davinci Resolve/Fusion for editing and composition.

Sometimes we only want VFX to be applied on single elements on the scene. But unfortunately with hundres of elements in the spine 'scene' it's difficult (and error prone) to create different isolated 'layer'-exports from Spine.
It would be nice if there would be something like a layering system inside of Spine so we could export tracks isolated and can combine them, after applying VFX, in other tools later in the workflow.

2) Merge animations

Another thing I just bumped into is that I needed one animation that was quite hard work and difficult to create. And I thought it would be extremely time consuming to create in Spine, because a lot of elements needed to be copied and all hand animated.
So I created a different animation in the file for it to export that as a single video on alpha and figured I would combine and loop it in the video editor later.

In the end the result seemed to be just fine in Spine already, but now I have the problem that all keyframes are made in the other animation object. And AFAIK there's no way to merge animations inside of Spine.

It would be nice to be able to merge the data of different animations to a new animation in Spine. That would make it possible to first tryout some animations on parts of the scene in a different animation holder instead of messing up the main scene. And when succesful we could then merge the animation date together with the other selected animation data (like the 'main' scene) so it's one animation to export to data or video.

**Comment by bencresty** [comment]
FYI I just wrote a nodejs script that merges requested animations from inside a Spine json file and adds this as a new 'merged' animation in the output.

It was actually easier than I thought; just did a deep merge on the animation object inside the json file to merge the animations to a new one and add that to the animations object. I assumed the creator of the animations knows what he/she's doing and doesn't animate the same properties in the different input animations, so it could perhaps use some degree of fine tuning to be used in weird situations and catching things like duplicates, but besides that on thoughtful use I think it's working fine and is pretty doable to just deep merge the animation object! :)

**Comment by NathanSweet** [comment]
Re: isolated export, could you organize your "layers" as separate skeletons? Then you could export a single skeleton. This won't work if you need to mix the draw order of your layers, but since you are exporting them as layers I don't think that is an issue.

Another option could be to use skins to group attachments, but a skeleton per layer seems to make more logical sense.

We have ideas for a selection system that could be related to export, but it's still early days for that.

Re: merging animations, if you have two animations within the same skeleton, you can copy keys from one to the other. If the animations are in different skeletons, then you cannot because the other skeleton has different bones/slots/etc. Even if they have the same names, we don't want to make such big assumptions for simple copy/paste. It would only work if everything needed was there (which quickly gets complicated with skins, etc) and we'd need to provide feedback if problems occurred.

While you can't do it with copy/paste, you can have Spine merge animations from different skeletons by using [Import project](http://esotericsoftware.com/spine-import#Project) and choosing `Animation`. Spine will match things up the best it can and provide a log of any problems encountered.

That can be used, for example, by teams that have a "master" project with all the bones/slots/attachments. A copy can be given to team members who only create animations, but do not create any new rigging (no new bones/slots/attachments). Later, the animations in the copies can be imported into the master project. To do new rigging, it needs to be done in the master project and new copies provided to the team.

**Comment by bencresty** [comment]
> Re: isolated export, could you organize your "layers" as separate skeletons? Then you could export a single skeleton. This won't work if you need to mix the draw order of your layers, but since you are exporting them as layers I don't think that is an issue.

In the end I needed draw order, both back and front of other parts, so merging animation data with a custom script was the only solution for that AFAIK. It would take forever to copy and paste all keyframes from one animation to another as it has a lot of properties animated. Or perhaps I am missing something? Need to try it out someday, but for this use case it wasn't a solution either way, because I'd liked to keep the animations seperate in own animation containers to keep working in seperate animations and not having everything animating all over the place while working on a single animation (and still show other items in the scene, but without movements), while everything still in the same scene.

I've never worked with several skeletons in a file, but AFAIK we can't show both at the same time? (I should check it out soon) If that's the case it wouldn't be a solution as we need to see everything at the same time, as it's all part of the same scene.

> Another option could be to use skins to group attachments, but a skeleton per layer seems to make more logical sense.

Sorry, that wouldn't be possible either as these are completely different items with no things incommon.

> We have ideas for a selection system that could be related to export, but it's still early days for that.

Great to read this!
I know Spine is mainly for games and I use the interactive aspect of it too. But also working for video and this would be extremely useful for creating character animations for video especially because we can then use VFX on seperate channels. Or, for instance, use all parts as seperate layers in 3d levels in Blender. Just to name a few use cases. That would be very powerful to keep things seperated, but make them powerful together with other software. That way Spine can focus on everything character animation, while VFX software can focus on all the effects and compositing, or interaction with other things in the video, like 3d. Looking forward to it!

> While you can't do it with copy/paste, you can have Spine merge animations from different skeletons by using [Import project](http://esotericsoftware.com/spine-import#Project) and choosing Animation. Spine will match things up the best it can and provide a log of any problems encountered.

Interesting! So you mean importing data is actually merging animation if you don't choose to create a new project? Cool! That's something I'm gonna check out. Great idea!

BTW thanks for all the reactions and active work on these issues! Very much appreciated!

**Comment by NathanSweet** [comment]
Spine has multiple skeletons visible at once. Give it a try with `ctrl+N` or duplicate your whole skeleton a few times and move them so you can see all at once. This is close to the "layers" system you mentioned. Skeletons can be dragged up/down in the tree to control which is on top. It only falls down when you need to intermingle the draw order -- a skeleton is either above or below all other skeletons. If you need to mix the draw order, then you need to build your stuff within a single skeleton. In that case you no longer have separate layers for export, since you are mixing the draw order.

> So you mean importing data is actually merging animation if you don't choose to create a new project? 

Not `Import Data`, I'm talking about `Import Project`. It can either bring a skeleton from another project (or the same project) into your current project, or bring animations from another project (or the same project) into an existing skeleton in your current project. Yep, it's doing the merge by looking at the names of bones, slots, etc.

> BTW thanks for all the reactions and active work on these issues! Very much appreciated!

Sure, no problem! The appreciation certainly goes both ways. You've brought a lot of good ideas and discussion lately and Spine is better for it, cheers! We try to be accommodating as much as possible, especially for changes that don't require major effor or redesigns or go against our reasoning for the decisions we've made. I'm happy to explain our thinking where we've done things purposefully, so I hope you don't get discouraged if we end up disagreeing on a point here or there. Our goal is to try to do what is best considering all of Spine's uses and workflows, tempered by the manpower we have available and hours in the day.

**Comment by bencresty** [comment]
Thanks, it wasn't really a solution for the project I was working on as it has overlapping parts both front and back, but I'll give it a try because it looks interesting for other use cases.

**Comment by bencresty** [comment]
@NathanSweet  Sorry for testing this so late, but I just did it;

Import Project looks like a very useful tool to know, to import a skeleton, but especially here to import animations where the same bones, attachments etc. are used to keyframe.

However, I tried it to merge animations and that doesn't seem to be possible. Even when I import a Spine file with an animation called `anim` in the current file where there's also an animation called `anim` it won't merge the two. When importing it creates an `anim2` next to the already existing animation.

This was just for your interest and perhaps for future reference, because since I'd created a tool to merge animations anyhow this feature isn't really needed in Spine here anymore. Could be nice to set a checkbox during import to merge animations with the same name instead of creating a copy, but if it's too much work and nobody else would use it at the moment you guys could probably use the costly time better I think!
